multichain-aggregator/multichain-aggregator-logic/src/repository/addresses.rs

```diff
@@ -79,20 +72,45 @@ where
         .apply_if(token_types, |q, token_types| {
             q.filter(Column::TokenType.is_in(token_types))
         })
-        .cursor_by((Column::Hash, Column::ChainId));
+        .apply_if(page_token, |q, page_token| {
+            q.filter(
+                Expr::tuple([
+                    Column::Hash.into_simple_expr(),
+                    Column::ChainId.into_simple_expr(),
+                ])
+                .gte(Expr::tuple([
+                    page_token.0.as_slice().into(),
+                    page_token.1.into(),
+                ])),
+            )
+        })
+        // Because of ORDER BY (primary_key) and LIMIT clauses, query planner chooses to use pk index
+        // instead of specialized indexes on filtered columns, which almost always results in a seqscan
+        // because in our case data is sparse (especially for text columns). To prevent it, we wrap
+        // the ordered columns in a COALESCE which makes query planner think it is an expression
+        // and disregards the primary key index.
+        .order_by_asc(Expr::cust_with_exprs(
+            "COALESCE($1)",
+            [Column::Hash.into_simple_expr()],
+        ))
+        .order_by_asc(Expr::cust_with_exprs(
+            "COALESCE($1)",
+            [Column::ChainId.into_simple_expr()],
+        ))
+        .limit(page_size + 1)
+        .all(db)
+        .await?;
 
-    if let Some(page_token) = page_token {
-        c.after((page_token.0.as_slice(), page_token.1));
+    match addresses.get(page_size as usize) {
+        Some(a) => Ok((
+            addresses[..page_size as usize].to_vec(),
+            Some((
+                AddressAlloy::try_from(a.hash.as_slice()).unwrap(),
```

Comment:
is the removed comment about unwrap no longer relevant?

Reply:
Accidentally deleted. Just restored it.